src/Riok.Mapperly/Descriptors/MappingBodyBuilder/NewInstanceObjectPropertyMappingBodyBuilder.cs

```diff
@@ -195,42 +195,88 @@ private static void BuildInitOnlyPropertyMappings(NewInstanceMappingBuilderConte
+                    return false;
+
+                skippedOptionalParam = true;
+                continue;
+            }
+
+            // nullability is handled inside the property mapping
+            var paramType = parameter.Type.WithNullableAnnotation(parameter.NullableAnnotation);
+            var delegateMapping = ctx.BuilderContext.FindMapping(sourcePath.MemberType, paramType)
+                ?? ctx.BuilderContext.FindOrBuildMapping(sourcePath.Member.Type.NonNullable(), paramType);
+            if (delegateMapping == null)
+            {
+                if (!parameter.IsOptional)
+                    return false;
+
+                skippedOptionalParam = true;
+                continue;
+            }
+
+            var propertyMapping = new NullPropertyMapping(delegateMapping, sourcePath, ctx.BuilderContext.GetNullFallbackValue(paramType));
+            var ctorMapping = new ConstructorParameterMapping(parameter, propertyMapping, skippedOptionalParam);
+            constructorParameterMappings.Add(ctorMapping);
             mappedTargetPropertyNames.Add(parameter.Name);
         }
 
         return true;
     }
+
+    private static bool TryFindConstructorParameterSourcePath(NewInstanceMappingBuilderContext ctx, IParameterSymbol parameter, [NotNullWhen(true)] out PropertyPath? sourcePath)
+    {
+        sourcePath = null;
+
+        if (ctx.PropertyConfigsByRootTargetName.TryGetValue(parameter.Name, out var propertyConfigs))
+        {
+            if (propertyConfigs.Count > 1)
+            {
+                ctx.BuilderContext.ReportDiagnostic(
+                    DiagnosticDescriptors.MultipleConfigurationsForConstructorArgument,
+                    parameter.Type,
+                    parameter.Name);
+            }
+
+            var propertyConfig = propertyConfigs.First();
+            if (propertyConfig.Target.Count > 1)
+            {
+                ctx.BuilderContext.ReportDiagnostic(
+                    DiagnosticDescriptors.ConstructorArgumentDoesNotSupportPaths,
+                    parameter.Type,
+                    string.Join(".", propertyConfig.Target));
+                return false;
+            }
+
+            if (!PropertyPath.TryFind(
+                ctx.Mapping.SourceType,
+                propertyConfig.Source,
+                out sourcePath))
+            {
+                ctx.BuilderContext.ReportDiagnostic(
+                    DiagnosticDescriptors.MappingSourcePropertyNotFound,
+                    propertyConfig.Source,
+                    ctx.Mapping.SourceType);
+                return false;
+            }
+
+            return true;
+        }
+
+        if (!PropertyPath.TryFind(
+            ctx.Mapping.SourceType,
+            MemberPathCandidateBuilder.BuildMemberPathCandidates(parameter.Name),
+            ctx.IgnoredSourcePropertyNames,
+            StringComparer.OrdinalIgnoreCase,
+            out sourcePath)
+        )
+            return false;

```

Comment:
If the if clause spans over multiple lines, this codebase usually uses braces for the body of the if. However this could be transformed in a direct return:
```csharp
return PropertyPath.TryFind(...
```